prov/util/src/util_srx.c

```diff
@@ -0,0 +1,1121 @@
+		if (rx_entry->peer_entry.context == context) {
+			slist_remove(queue, item, prev);
+			util_cancel_entry(srx, flags, rx_entry);
+			srx->lock_release(&srx->lock);
+			return true;
+		}
+	}
+	srx->lock_release(&srx->lock);
+	return false;
+}
+
+static int util_cancel_src(struct ofi_dyn_arr *arr, void *list, void *context)
+{
+	struct util_srx_ctx *srx;
+	struct slist *queue = list;
+	uint64_t flags;
+
+	srx = container_of(arr, struct util_srx_ctx, src_trecv_queues);
+
+	flags = arr == &srx->src_trecv_queues ?
+		FI_TAGGED | FI_RECV : FI_MSG | FI_RECV;
+
+	return (int) util_cancel_recv(srx, queue, flags, context);
+}
+
+static ssize_t util_srx_cancel(fid_t ep_fid, void *context)
+{
+	struct util_srx_ctx *srx;
+
+	srx = container_of(ep_fid, struct util_srx_ctx, peer_srx.ep_fid);
+
+	if (util_cancel_recv(srx, &srx->tag_queue, FI_TAGGED | FI_RECV,
+			     context))
+		return 0;
+
+	if (util_cancel_recv(srx, &srx->msg_queue, FI_MSG | FI_RECV,
+			     context))
+		return 0;
+
+	if (ofi_array_iter(&srx->src_trecv_queues, context, util_cancel_src))
+		return 0;
+
+	if (ofi_array_iter(&srx->src_recv_queues, context, util_cancel_src))
+		return 0;
+
+	return 0;
+}
+
+static int util_srx_getopt(fid_t fid, int level, int optname,
+		           void *optval, size_t *optlen)
+{
+	struct util_srx_ctx *srx =
+		container_of(fid, struct util_srx_ctx, peer_srx.ep_fid.fid);
+
+	if ((level != FI_OPT_ENDPOINT) || (optname != FI_OPT_MIN_MULTI_RECV))
+		return -FI_ENOPROTOOPT;
+
+	*(size_t *)optval = srx->min_multi_recv_size;
+	*optlen = sizeof(size_t);
+
+	return FI_SUCCESS;
+}
+
+static int util_srx_setopt(fid_t fid, int level, int optname,
+		           const void *optval, size_t optlen)
+{
+	struct util_srx_ctx *srx =
+		container_of(fid, struct util_srx_ctx, peer_srx.ep_fid.fid);
+
+	if ((level != FI_OPT_ENDPOINT) || (optname != FI_OPT_MIN_MULTI_RECV))
+		return -FI_ENOPROTOOPT;
+
+	srx->min_multi_recv_size = *(size_t *)optval;
+
+	return FI_SUCCESS;
+}
+
+static struct fi_ops_ep util_srx_ops = {
+	.size = sizeof(struct fi_ops_ep),
+	.cancel = util_srx_cancel,
+	.getopt = util_srx_getopt,
+	.setopt = util_srx_setopt,
+	.tx_ctx = fi_no_tx_ctx,
+	.rx_ctx = fi_no_rx_ctx,
+	.rx_size_left = fi_no_rx_size_left,
+	.tx_size_left = fi_no_tx_size_left,
+};
+
+static void util_rx_entry_init_fn(struct ofi_bufpool_region *region, void *buf)
+{
+	struct util_rx_entry *rx_entry = (struct util_rx_entry *) buf;
+	struct util_srx_ctx *srx = (struct util_srx_ctx *)
+					region->pool->attr.context;
+
+	rx_entry->peer_entry.iov = util_srx_iov(rx_entry);
+	rx_entry->peer_entry.desc = util_srx_desc(srx, rx_entry);
+}
+
+static void util_srx_init_slist(struct ofi_dyn_arr *arr, void *item)
+{
+	slist_init((struct slist *) item);
+}
+
+static void util_srx_init_dlist(struct ofi_dyn_arr *arr, void *item)
+{
+	dlist_init((struct dlist_entry *) item);
+}
+
+int util_ep_srx_context(struct util_domain *domain, size_t rx_size,
+			size_t iov_limit, size_t default_min_mr,
```

Comment:
nit: I realize `default_min_mr` is actually for multi recv, but the name makes me think of it's memory registration